jdk/test/sun/nio/cs/TestCharsetMapping.java

```diff
@@ -0,0 +1,640 @@
+            this.csName = csName;
+            this.clzName = clzName;
+            if (csName.endsWith("_Solaris") ||
+                csName.endsWith("_MS5022X") ||
+                csName.endsWith("_MS932")) {
+                isInternal = true;
+            }
+        }
+
+        private Entry parse(Matcher m) {
+            Entry e = new Entry();
+            e.bb = Long.parseLong(m.group(G_BS), 16);
+            if (e.bb < 0x100)
+                e.bs = new byte[] { (byte)e.bb };
+            else
+                e.bs = parseBytes(m.group(G_BS));
+            e.cp = Integer.parseInt(m.group(G_CP), 16);
+            if (G_CP2 <= m.groupCount() && m.group(G_CP2) != null)
+               e.cp2 = Integer.parseInt(m.group(G_CP2), 16);
+            else
+               e.cp2 = 0;
+            return e;
+        }
+
+        boolean loadMappings(Path dir) throws IOException {
+            // xxx.map
+            Path path = dir.resolve(clzName + ".map");
+            if (!Files.exists(path)) {
+                return false;
+            }
+            Matcher m = ptn.matcher("");
+            mappings = Files.lines(path)
+                .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                .map(ln -> parse(m))
+                .filter(e -> e.cp != UNMAPPABLE)  // non-mapping
+                .collect(Collectors.toList());
+            // xxx.nr
+            path = dir.resolve(clzName + ".nr");
+            if (Files.exists(path)) {
+                nr = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.bb, Function.identity()));
+            }
+            // xxx.c2b
+            path = dir.resolve(clzName + ".c2b");
+            if (Files.exists(path)) {
+                c2b = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.cp, Function.identity()));
+            }
+            return true;
+        }
+    }
+
+    private static Set<CharsetInfo> charsets(Path cslist, boolean sbcs) throws IOException {
+        Set<CharsetInfo> charsets = new LinkedHashSet<>();
+        Iterator<String> itr = Files.readAllLines(cslist).iterator();
+
+        while (itr.hasNext()) {
+            String line = itr.next();
+            if (line.startsWith("#") || line.length() == 0) {
+                continue;
+            }
+            String[] tokens = line.split("\\s+");
+            if (tokens.length < 5) {
+                continue;
+            }
+            CharsetInfo cs = new CharsetInfo(tokens[1], tokens[0]);
+            cs.hisName = tokens[2];
+            cs.pkgName = tokens[4];
+            if (sbcs) {
+                cs.type = "sbcs";
+            } else {
+                cs.type = tokens[3];
+            }
```

Comment:
Mentally parsing:

Format 1, dbcs:
clzName  csName     hisName  dbtype    pkg               ascii   b1min  b1max  b2min b2max

hisname = 2
pkgName = 4
type = sbcs

Format 2, sbcs and extsbcs:
clzName	csName		hisName		containASCII	pkg

hisname = 2
pkgName = 4
type = sbcs

Okay, checks out.

Reply:
Could you add a small comment here, listing these two formats for easier code reading:

```
// dbcs format
// clzName csName hisName dbtype pkg ascii b1min b1max b2min b2max
// sbcs format
//  clzName csName hisName containASCII pkg
```

?

Reply:
Maybe also some sanity tests:
- pkgname to start with "sun.nio"
- type one of "basic ebcdic euc_sim dbcsonly sbcs"

Reply:
The comment is a good idea. I was looking at that line in the mapping file as I was writing the code, and it would be good to have it in there directly.

I think the sanity tests are something we should add in mainline first and backport, as both would apply there too (though mainline has a few more types).